viewing_party/party.py

```diff
@@ -1,23 +1,229 @@
+        return []
+    if not user_data["friends"]:
+        return user_data["watched"]
+    user_unique_watched = []
+    for movie in user_data["watched"]:
+        is_unique = True        
+        for friend in user_data["friends"]:
+            if not friend["watched"]:
+                continue
+            if movie in friend["watched"]:
+                is_unique = False
+                break
+        if is_unique:
+            user_unique_watched.append(movie) 
+    return user_unique_watched
+
+# TC: O(f * m * (n + k))
+# SC: O(k)
+# f -- number of friends
+# m -- average number of movies watched by each friend
+# n -- length of user['watched']
+# k -- length of friends_unque_watched_movie
+
+def get_friends_unique_watched(user_data):
+    """
+    Get the movies which the friends have watched, but user has not.
+    Parameter: user_data(dict)
+    Return: List of friends' unique movies(list)
+    """
+    friends_unique_watched = []
+    if not user_data["friends"]:
+        return []
+    
+    for friend in user_data["friends"]:
+        if not friend["watched"]:
+            continue
+        for movie in friend["watched"]:
+            if not user_data["watched"] or \
+                (movie not in user_data["watched"] and 
+                 movie not in friends_unique_watched):
+                friends_unique_watched.append(movie)
+    return friends_unique_watched
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+# TC: O(f * m * (n + k)) + O(k * l)
+# SC: O(k)
+# f -- number of friends
+# m -- average number of movies watched by each friend
+# n -- length of user['watched']
+# k -- length of friends_unque_watched_movie
+# l -- length of user_data's subscription list;
+def get_available_recs(user_data):
+    """
+    Get a list of reccomended movies that that the user has not watched but a 
+    friend has. 
+    This movie must be hosted on one of the user's subscription services.
+    Parameters: user_data(dict)
+    Return: list of reccomended movies(list)
+    """
+    movies = get_friends_unique_watched(user_data)
+    return [movie for movie in movies 
+            if movie["host"] in user_data["subscriptions"]]
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+# For the following two functions: 
+# TC and SC is the same as "get_unique_watched()" as it is the main 
+# part of operations
+# TC: O(n*f*m)
+# n -- length of user['watched']
+# f -- number of friends
+# m -- average number of movies watched by each friend
+# SC: O(k) length of unique movies watched by user, worst case O(n)
+def get_new_rec_by_genre(user_data):
+    """
+    Get movies from the user's most frequently watched genre which the user has 
+    not watched, but a friend has watched.
+    Parameters: user_data(dict)
+    Return: List of reccomended movies(list)
+    """

```

Comment:
In this case and a couple others, the docstring you've included is actually longer than the function itself! In these cases, think about either shortening the docstring or removing it altogether!